community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/codegen/ir/Instruction.scala

```diff
@@ -22,48 +22,44 @@ package org.neo4j.cypher.internal.compiler.v2_3.codegen.ir
 import org.neo4j.cypher.internal.compiler.v2_3.codegen.ExceptionCodeGen
 
 trait Instruction {
+
   // Actual code produced by element
   def generateCode(): String
 
-  final def importedClasses() = allLeafs.flatMap(_._importedClasses()).toSet
+  //Initialises necessary data-structures. Is inserted at the top of the generated method
+  def generateInit(): String
 
   //generate class level members
   def members(): String
 
-  //Initialises necessary data-structures. Is inserted at the top of the generated method
-  def generateInit(): String
+  protected def children: Seq[Instruction]
 
-  def methods: Seq[Method] = {
-    (allLeafs :+ this).flatMap(_._method)
+  private def treeView: Seq[Instruction] = {
+    children.foldLeft(this +: children) { case (acc, child) => acc ++ child.treeView
```

Comment:
Hm, but we still need the implementing class to be included. Either we do it like this, with duplicates (which works since we're using sets elsewhere, but I agree is a bit shaky), or we change this to use a `Set` rather than a `Seq`, or use one 'header' function to add `this` and one recursive function for the rest. What do you think?